libs/cache/CMakeLists.txt

```diff
@@ -5,150 +5,64 @@
-    ${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility/hpx/util/cache/entries/entry.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility/hpx/util/cache/entries/fifo_entry.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility/hpx/util/cache/entries/lfu_entry.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility/hpx/util/cache/entries/lru_entry.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility/hpx/util/cache/entries/size_entry.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility/hpx/util/cache/policies/always.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility/hpx/util/cache/statistics/local_full_statistics.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility/hpx/util/cache/statistics/local_statistics.hpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility/hpx/util/cache/statistics/no_statistics.hpp
+if(HPX_${upper_module_name}_WITH_COMPATIBILITY_HEADERS)
+  set(${module_name}_compat_headers
+    ${compat_header_root}/hpx/util/${module_name}/local_cache.hpp
+    ${compat_header_root}/hpx/util/${module_name}/lru_cache.hpp
+    ${compat_header_root}/hpx/util/${module_name}/entries/entry.hpp
+    ${compat_header_root}/hpx/util/${module_name}/entries/fifo_entry.hpp
+    ${compat_header_root}/hpx/util/${module_name}/entries/lfu_entry.hpp
+    ${compat_header_root}/hpx/util/${module_name}/entries/lru_entry.hpp
+    ${compat_header_root}/hpx/util/${module_name}/entries/size_entry.hpp
+    ${compat_header_root}/hpx/util/${module_name}/policies/always.hpp
+    ${compat_header_root}/hpx/util/${module_name}/statistics/local_full_statistics.hpp
+    ${compat_header_root}/hpx/util/${module_name}/statistics/local_statistics.hpp
+    ${compat_header_root}/hpx/util/${module_name}/statistics/no_statistics.hpp
   )
 endif()
 
-set(sources
-  ${CMAKE_CURRENT_SOURCE_DIR}/src/cache.cpp
+set(${module_name}_sources
+  ${source_root}/cache.cpp
 )
 
-add_library(hpx_cache STATIC ${sources} ${headers} ${compat_headers})
-
-target_link_libraries(hpx_cache hpx_config)
-target_include_directories(hpx_cache PUBLIC
-  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
-  $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}>)
-
-if(HPX_CACHE_WITH_COMPATIBILITY_HEADERS)
-  target_include_directories(hpx_cache PUBLIC
-    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility>)
-endif()
-
-target_compile_definitions(hpx_cache PRIVATE
-  $<$<CONFIG:Debug>:DEBUG>
-  $<$<CONFIG:Debug>:_DEBUG>
-  HPX_MODULE_EXPORTS
+include(HPX_AddModule)
+add_hpx_module(${module_name}
+    DEPRECATION_WARNINGS
+    COMPATIBILITY_HEADERS
+    FOLDER ${CMAKE_CURRENT_SOURCE_DIR}
+    SOURCE_ROOT ${source_root}
+    HEADER_ROOT ${header_root}
+    COMPAT_HEADER_ROOT ${compat_header_root}
+    SOURCES ${${module_name}_sources}
+    HEADERS ${${module_name}_headers}
+    COMPAT_HEADERS ${${module_name}_compat_headers}
+    DEPENDENCIES hpx_config
+    CMAKE_SUBDIRS examples tests
 )
```

Comment:
Should we add AUTOGLOB capabilities to this function (see add_hpx_library() and friends)? I know that generally people discourage the use of file system globbing for cmake projects as it might break cmake's own dependency tracking. It might however simplify creating modules significantly as the user wouldn't have to explicitly list all header and source files...

Reply:
I don't know cause keeping the headers like this makes it safer, and we don't create modules that often, what do you think ?

Reply:
Yah, it's probably not a big issue - definitely not a showstopper for this PR.